### PR TITLE
[asset-backfill] Fix edge case when grouping runs with backfill policies and partition mappings.

### PR DESCRIPTION
**Pull request by OwenKephart**
## Summary & Motivation

This PR refactors how we detect "simple" partition mappings, and limits the scope of when we can group assets into runs in that case.

In terms of "simple" partition mappings, we now explicitly check for Identity / TimeWindow partition mappings with no offsets. Previously, we were looking for the case of "has a single upstream parent partition that is the same as the child partition", which certainly sounds like a "simple partition mapping" but actually does not cover all potential cases.

In particular, imagine you have two daily-partitioned assets, with the downstream asset depending on the previous N days of the upstream. In this case, the first partition of the downstream asset only has a single parent, and that partition is the same as the child partition, even though this isn't a "simple" partition mapping! This meant it would be erroneously allowed to run with its parent.

For the scope of when we group assets together, there were too many weird cases to consider here and so instead I just reworked this logic entirely to only allow weird partition mappings to run together if:

- The exact same set of upstream / downstream partitions are targeted
- All upstream partitions are set to be materialized on this tick

This means that we can be confident that all downstream partitions will be requested this tick, and that they will be the same exact partitions (meaning they will be grouped together into the same run)

## How I Tested These Changes

Test failed before, passes now

### CODE REVIEW

File: python_modules/dagster/dagster/_core/execution/asset_backfill.py [hunk 1]
```diff
@@ -1376,7 +1379,7 @@ def should_backfill_atomic_asset_partitions_unit(
                     (
```
Comment (clairelin135) [inline python_modules/dagster/dagster/_core/execution/asset_backfill.py:1379]: Re the backfill issue earlier today:

We probably could remove this block as well as `has_partition_mapping_safe_backfill_policy` and instead allow a candidate to run with the parent if:
- They have the same backfill policy
- There is a simple 1:1 partition mapping where the parent partition key is the same as the child's. I think this only applies to the `IdentityPartitionMapping` and the default `TimeWindowPartitionsMapping`

which should fix this multi-asset issue too

**Review by clairelin135 — COMMENTED** [review]
The fix here looks good to me, if you did want to merge it in separate from the backfill issue today.

Included an inline note in case if you wanted to edit this PR to handle the backfill fix.

File: python_modules/dagster/dagster/_core/execution/asset_backfill.py (hunk 1 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster/_core/execution/asset_backfill.py]: So, spent a good amount of time on this and agree that the existing state of `has_partition_mapping_safe_backfill_policy` was problematic. However, there is actually a good(ish) reason to add an additional path to being grouped with your parents beyond those first two.

Imagine you have time-partitioned two assets with a fancy lookback partition mapping between them. However, you're actually doing a backfill that targets all partitions of each of those assets. The optimal solution here is to materialize this all in a single run that targets both of these assets (and all partitions for each of them). Even though there's some fancy mapping, we know that all partitions of A will materialize before any B partitions start.

That's what I encapsulated in the new-and-improved additional block of logic. This is much more specific / targeted than before, and should be a lot safer than the previous behavior (which was itself an improvement on previous behavior).

File: python_modules/dagster/dagster_tests/core_tests/execution_tests/test_asset_backfill_with_backfill_policies.py [hunk 2]
```diff
@@ -574,7 +575,8 @@ def asset1() -> None:
 
 def test_assets_backfill_with_partition_mapping():
     daily_partitions_def: DailyPartitionsDefinition = DailyPartitionsDefinition("2023-01-01")
-    time_now = pendulum.now("UTC")
+    # time at which there will be an identical set of partitions for the downstream asset
+    test_time = pendulum.parse("2023-03-04T00:00:00", tz="UTC")
```
Comment (OwenKephart) [inline python_modules/dagster/dagster_tests/core_tests/execution_tests/test_asset_backfill_with_backfill_policies.py:579]: these tests experience a new behavior if you just stick with the current time. previously, the first run emitted would be:

- both A and B, 2023-03-01 to 2023-03-03

We'd still need to materialize 2023-03-04 to 2023-06 for B, but these would need to wait for their parent partitions of A to materialize.

This is actually logically correct behavior in this instance, but risky to implement across the board. It's hard to know in general that the exact same set of partitions will be emitted for B as were requested for A, as when we're calculating if a single partition of B should be materialized, we don't have access to the decisions for all other partitions of B.

Now, we're more conservative and only consider cases where the target set of B is identical to the target set of A, which in this case can happen if  those partitions of B which depend on older partitions of A don't yet exist.

In the previous case (where they do exist), the behavior is still reasonable (assuming all partitions of A are materialized within the same tick, we'll just end up kicking one run for all A partitions and one run for all B partitions, resulting in the same total number of runs as the previous behavior).

Comment (clairelin135) [reply python_modules/dagster/dagster_tests/core_tests/execution_tests/test_asset_backfill_with_backfill_policies.py:579]: Thanks for explaining, this sounds reasonable to me.

> In the previous case (where they do exist), the behavior is still reasonable (assuming all partitions of A are materialized within the same tick, we'll just end up kicking one run for all A partitions and one run for all B partitions, resulting in the same total number of runs as the previous behavior).

Thoughts on adding a test with this case to document the expected behavior?

File: python_modules/dagster/dagster/_core/execution/asset_backfill.py [hunk 3]
```diff
@@ -1352,36 +1353,60 @@ def should_backfill_atomic_asset_partitions_unit(
...
-                        parent.partition_key == candidate.partition_key
-                        and len(parent_partitions_result.parent_partitions) == 1
+                    # if there is a simple mapping between the parent and the child, then
+                    # with the parent
+                    has_identity_partition_mapping
+                    # if there is not a simple mapping, we can only materialize this asset with its
+                    # parent if...
+                    or (
+                        # there is a backfill policy for the parent
+                        parent_backfill_policy is not None
```
Comment (clairelin135) [inline python_modules/dagster/dagster/_core/execution/asset_backfill.py:1394]: I believe we [don't allow assets with different backfill policies to be targeted in the same run](https://github.com/dagster-io/dagster/blob/claire/asset-overview-page/python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py#L436-L466)...

So if we had upstream A with `BackfillPolicy.multi_run(30)` and downstream B with `BackfillPolicy.multi_run(60)` and an identity partition mapping in between, I believe the current implementation would kick off both assets in separate runs.

I think we might need to make `can_run_with_parent` return False if the backfill policies are not the same? Probably helpful to add a test with this scenario too.

**Review by clairelin135 — COMMENTED** [review]
Thanks for adding a fix for this, this is looking good to me. Added another edge case for your consideration in the inline comments.

File: python_modules/dagster/dagster/_core/execution/asset_backfill.py (hunk 1 above)
Comment (clairelin135) [reply python_modules/dagster/dagster/_core/execution/asset_backfill.py]: Got it -- that makes sense. The new block of logic that handles the "all partitions selected" case looks good to me.

File: python_modules/dagster/dagster_tests/core_tests/execution_tests/test_asset_backfill_with_backfill_policies.py (hunk 2 above)
Comment (OwenKephart) [reply python_modules/dagster/dagster_tests/core_tests/execution_tests/test_asset_backfill_with_backfill_policies.py:579]: good call, will add

File: python_modules/dagster/dagster/_core/execution/asset_backfill.py (hunk 3 above)
Comment (clairelin135) [reply python_modules/dagster/dagster/_core/execution/asset_backfill.py:1394]: ohhhhh gotcha, thanks!

**Comment by OwenKephart** [comment]
@clairelin135 refactored all this "can run with parent" logic into its own function for two reasons:

1. readability (this thing is a bunch of logic)
2. perf (this involves a lot of checking stuff, so it's nice to not even call into this function if (e.g.) the parent is not in the target subset.

**Review by clairelin135 — APPROVED** [review]
LGTM!

**Pull request merged**